src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -168,69 +181,96 @@ protected IRequestExecutionQueue<TRequestContext> GetRequestExecutionQueue()
+        return LanguageServerConstants.DefaultLanguageName;
+    }
+
     private sealed class DelegatingEntryPoint
     {
         private readonly string _method;
+        private readonly Lazy<ImmutableDictionary<string, (MethodInfo MethodInfo, RequestHandlerMetadata Metadata)>> _languageEntryPoint;
         private readonly AbstractLanguageServer<TRequestContext> _target;
 
-        private static readonly MethodInfo s_entryPointMethod = typeof(DelegatingEntryPoint).GetMethod(nameof(EntryPointAsync))!;
-        private static readonly MethodInfo s_parameterlessEntryPointMethod = typeof(DelegatingEntryPoint).GetMethod(nameof(ParameterlessEntryPointAsync))!;
-        private static readonly MethodInfo s_notificationMethod = typeof(DelegatingEntryPoint).GetMethod(nameof(NotificationEntryPointAsync))!;
-        private static readonly MethodInfo s_parameterlessNotificationMethod = typeof(DelegatingEntryPoint).GetMethod(nameof(ParameterlessNotificationEntryPointAsync))!;
+        private static readonly MethodInfo s_queueExecuteAsyncMethod = typeof(RequestExecutionQueue<TRequestContext>).GetMethod(nameof(RequestExecutionQueue<TRequestContext>.ExecuteAsync))!;
 
-        public DelegatingEntryPoint(string method, AbstractLanguageServer<TRequestContext> target)
+        public DelegatingEntryPoint(string method, AbstractLanguageServer<TRequestContext> target, IHandlerProvider handlerProvider)
         {
             _method = method;
             _target = target;
-        }
-
-        public static MethodInfo GetMethodInstantiation(Type? requestType, Type? responseType)
-            => (requestType, responseType) switch
+            _languageEntryPoint = new Lazy<ImmutableDictionary<string, (MethodInfo, RequestHandlerMetadata)>>(() =>
             {
-                (requestType: not null, responseType: not null) => s_entryPointMethod.MakeGenericMethod(requestType, responseType),
-                (requestType: null, responseType: not null) => s_parameterlessEntryPointMethod.MakeGenericMethod(responseType),
-                (requestType: not null, responseType: null) => s_notificationMethod.MakeGenericMethod(requestType),
-                (requestType: null, responseType: null) => s_parameterlessNotificationMethod,
-            };
-
-        public async Task NotificationEntryPointAsync<TRequest>(TRequest request, CancellationToken cancellationToken) where TRequest : class
-        {
-            var queue = _target.GetRequestExecutionQueue();
-            var lspServices = _target.GetLspServices();
-
-            _ = await queue.ExecuteAsync<TRequest, NoValue>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+                var handlerEntryPoints = new Dictionary<string, (MethodInfo, RequestHandlerMetadata)>();
+                foreach (var metadata in handlerProvider
+                    .GetRegisteredMethods()
+                    .Where(m => m.MethodName == method))
```

Comment:
>.Where(m => m.MethodName == method)) [](http://example.com/codeflow?start=20&length=36)

Almost feels like IHandlerProvider should have a GetMethodHandlers instead of GetMethodHandler